pkg/tsdb/prometheus/promclient/cache.go

```diff
@@ -0,0 +1,55 @@
+package promclient
+
+import (
+	lru "github.com/hashicorp/golang-lru"
+	apiv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+const (
+	noPassThrough = "no-pass-through"
+)
+
+type ProviderCache struct {
+	provider promClientProvider
+	cache    *lru.Cache
+	jsonData JsonData
+}
+
+type promClientProvider interface {
+	GetClient(map[string]string) (apiv1.API, error)
+}
+
+func NewProviderCache(p promClientProvider, jd JsonData) (*ProviderCache, error) {
+	cache, err := lru.New(500)
```

Comment:
This is a guess. Are there any thoughts as to how many clients we may want to cache at a time?

Reply:
am i understanding this correctly?

<img src="https://render.githubusercontent.com/render/math?math=$\frac{U}{P} = 500">

* `U` is the number of grafana users
* `P` is the number of configured prometheus instances in grafana with oauth forwarding enabled

so this could support caching 500 users with oauth forwarding turned on for 1 configured prometheus data source, 250 users for 2 prometheus data sources, ~166 users for 3, etc. does that sound right?

if so, maybe it should be a config option?

Reply:
The provider cache is created in `newInstanceSettings` here: https://github.com/grafana/grafana/blob/1d1cc25f543c6c6c0a4cfc12c6c1c1e36dfe89d4/pkg/tsdb/prometheus/prometheus.go#L59-L80

That means a new cache is created per prometheus data source. We're looking at 500 oauth tokens per prometheus data source

Reply:
@kminehart @marefr @radiohead any thought on how many clients we may want to cache at a time? 
 
 In https://github.com/grafana/grafana/pull/40349 because of incident, we have removed creating of client for each query and moved it to `newInstanceSettings`. We need to change it again for Oauth token forwarding. This means that we need to create a client per oauth-token. This PR also includes an lru cache of clients based on the passed auth headers.

Reply:
I looked at it for a chunk of the morning and I _think_ the cache size is fine. based on the sketchiest math: I think an upper bound for someplace with ~900 prometheus instances that all saturate their caches is looking at ~1GB at the largest for -- what I assume -- will already be a large installation.